pkg/generators/git_test.go

```diff
@@ -74,26 +95,101 @@ func TestGitGenerateParams(t *testing.T) {
 		{
 			name: "handles empty response from repo server",
 			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{"*"}},
+			files: nil,
 			repoApps: []string{},
+			repoPaths: []string{},
+			repoFileContents: nil,
 			repoError: nil,
 			expected: []map[string]string{},
 			expectedError:nil,
 		},
 		{
 			name: "handles error from repo server",
 			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{"*"}},
+			files: nil,
+			repoApps: []string{},
+			repoPaths: []string{},
+			repoFileContents: nil,
+			repoError: fmt.Errorf("error"),
+			expected: []map[string]string{},
+			expectedError: fmt.Errorf("error"),
+		},
+		{
+			name: "handles error from repo server",
+			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{"*"}},
+			files: nil,
 			repoApps: []string{},
+			repoPaths: []string{},
+			repoFileContents: nil,
 			repoError: fmt.Errorf("error"),
 			expected: []map[string]string{},
 			expectedError: fmt.Errorf("error"),
 		},
+		{
+			name: "happy flow: create params from git files",
+			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{"*"}},
+			files: nil,
+			repoApps: []string{},
+			repoPaths: []string{
+				"cluster-config/production/config.json",
+				"cluster-config/staging/config.json",
+			},
+			repoFileContents: map[string][]byte{
+				"cluster-config/production/config.json": []byte(`{
+   "cluster": {
+       "owner": "john.doe@example.com",
+       "name": "production",
+       "address": "https://kubernetes.default.svc"
+   },
+   "key1": "val1",
+   "key2": {
+       "key2_1": "val2_1",
+       "key2_2": {
+           "key2_2_1": "val2_2_1"
+       }
+   }
+}`),
+				"cluster-config/staging/config.json": []byte(`{
+   "cluster": {
+       "owner": "foo.bar@example.com",
+       "name": "staging",
+       "address": "https://kubernetes.default.svc"
+   }
+}`),
+			},
+			repoError: nil,
+			expected: []map[string]string{
+				{
+					"cluster.owner": "john.doe@example.com",
+					"cluster.name": "production",
+					"cluster.address": "https://kubernetes.default.svc",
+					"key1": "val1",
+					"key2.key2_1": "val2_1",
+					"key2.key2_2.key2_2_1": "val2_2_1",
+				},
+				{
+					"cluster.owner": "foo.bar@example.com",
+					"cluster.name": "staging",
+					"cluster.address": "https://kubernetes.default.svc",
+				},
+			},
+			expectedError: nil,
+		},
 	}
 
 	for _, c := range cases {
 		cc := c
 		t.Run(cc.name, func(t *testing.T) {
 			argoCDServiceMock := argoCDServiceMock{}
-			argoCDServiceMock.On("GetApps", mock.Anything, mock.Anything, mock.Anything).Return(c.repoApps, c.repoError)
+			mockCalls := argoCDServiceMock.
```

Comment:
@OmerKahani I started writing test cases for git file generator scenario, however I am not able to use `testify.Mock` properly. What I am trying to do is setup multiple expectations per test case:

- One for `GetPaths`
- Many for `GetFileContent`

I tried to use Mock chaining as mentioned here: https://godoc.org/github.com/stretchr/testify/mock#Call.On, but still could not get it working. The test is failing with error below:

```
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

time="2020-11-18T21:58:55+05:30" level=info msg="applications result from the repo service" allAps="[app1 app2 p1/app3]" repoURL=RepoURL revision=Revision total=3
time="2020-11-18T21:58:55+05:30" level=info msg="applications result from the repo service" allAps="[app1 p1/app2 p1/p2/app3 p1/p2/p3/app4]" repoURL=RepoURL revision=Revision total=4
time="2020-11-18T21:58:55+05:30" level=info msg="applications result from the repo service" allAps="[]" repoURL=RepoURL revision=Revision total=0
--- FAIL: TestGitGenerateParams (0.00s)
    --- FAIL: TestGitGenerateParams/happy_flow_-_created_apps (0.00s)
        git_test.go:220: PASS:  GetApps(string,string,string)
        git_test.go:220: FAIL:  GetPaths(string,string,string,string)
                        at: [git_test.go:186]
        git_test.go:220: FAIL: 1 out of 2 expectation(s) were met.
                The code you are testing needs to make 1 more call(s).
                at: [git_test.go:220]
    --- FAIL: TestGitGenerateParams/It_filters_application_according_to_the_paths (0.00s)
        git_test.go:220: PASS:  GetApps(string,string,string)
        git_test.go:220: FAIL:  GetPaths(string,string,string,string)
                        at: [git_test.go:186]
        git_test.go:220: FAIL: 1 out of 2 expectation(s) were met.
                The code you are testing needs to make 1 more call(s).
                at: [git_test.go:220]
    --- FAIL: TestGitGenerateParams/handles_empty_response_from_repo_server (0.00s)
        git_test.go:220: PASS:  GetApps(string,string,string)
        git_test.go:220: FAIL:  GetPaths(string,string,string,string)
                        at: [git_test.go:186]
        git_test.go:220: FAIL: 1 out of 2 expectation(s) were met.
                The code you are testing needs to make 1 more call(s).
                at: [git_test.go:220]
    --- FAIL: TestGitGenerateParams/handles_error_from_repo_server (0.00s)
        git_test.go:220: PASS:  GetApps(string,string,string)
        git_test.go:220: FAIL:  GetPaths(string,string,string,string)
                        at: [git_test.go:186]
        git_test.go:220: FAIL: 1 out of 2 expectation(s) were met.
                The code you are testing needs to make 1 more call(s).
                at: [git_test.go:220]
    --- FAIL: TestGitGenerateParams/handles_error_from_repo_server#01 (0.00s)
        git_test.go:220: PASS:  GetApps(string,string,string)
        git_test.go:220: FAIL:  GetPaths(string,string,string,string)
                        at: [git_test.go:186]
        git_test.go:220: FAIL: 1 out of 2 expectation(s) were met.
                The code you are testing needs to make 1 more call(s).
                at: [git_test.go:220]
FAIL
FAIL    command-line-arguments  1.222s
FAIL

```

Removing or commenting out the new `ArgoCDServiceMock.On` calls allows the test to pass.

Reply:
@dgoodwin I need some help here.

Reply:
@OmerKahani ^^

Reply:
Hi, sorry for the delay it was a busy week.

The problem is with line 186: https://github.com/argoproj-labs/applicationset/pull/45/files/1f6c6234c6514ba4d0fbf5693699c3c5ece966cc#diff-c1129e2ddfe8072c3267b62d8ac75bcd5dfa92143c40c478527853c08752b7eeR186

I think it will be best to create a new TestGitGenerateParams so there will be two functions:
TestGitGenerateParamsFromDirectory
TestGitGenerateParamsFromFile

everyone will have its own test cases and mock only one of the functions in the argocd service

Does it sound reasonable to you?

Reply:
Yes. Created separate test function for testing Git file generator.